### PR TITLE
replace app/victoria-metrics/main_test with apptest

### PR DESCRIPTION
**Pull request by jiekun**
### Describe Your Changes

https://github.com/VictoriaMetrics/VictoriaMetrics/issues/8466

This pull request:
1. Add `GraphiteWrite`, `CSVImport`, `OpenTSDBHTTPImport` data import methods for vmsingle and vminsert.
2. Add TCP `Write` method to `apptest.Client` to make it capable of writing data in TCP-based protocol.
3. Add test cases:
    1. for new import methods.
    2. for corner cases placed under `app/victoria-metrics/main_test`.
4. Removed all test cases under `app/victoria-metrics/main_test`.

### Checklist

The following checks are **mandatory**:

- [x] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

**Comment by jiekun** [comment]
Since it contains many test cases, it's recommended to review the change outside `apptest/tests` first.

File: apptest/client.go [hunk 1]
```diff
@@ -76,6 +75,25 @@ func (c *Client) do(t *testing.T, method, url, contentType string, data []byte)
...
+		_ = conn.Close()
+	}()
+
+	d := strings.Join(data, "\n")
+	n, err := conn.Write([]byte(d))
+	if err != nil {
+		t.Fatalf("cannot write %d bytes to %s: %s", len(data), address, err)
+	}
+	if n != len(d) {
+		panic(fmt.Errorf("BUG: conn.Write() returned unexpected number of written bytes to %s; got %d; want %d", address, n, len(data)))
```
Comment (rtm0) [inline apptest/client.go:93]: Just `t.Fatalf` this will stop the test immediately.

File: apptest/client.go [hunk 2]
```diff
@@ -76,6 +75,25 @@ func (c *Client) do(t *testing.T, method, url, contentType string, data []byte)
...
+	defer func() {
+		_ = conn.Close()
+	}()
+
+	d := strings.Join(data, "\n")
+	n, err := conn.Write([]byte(d))
+	if err != nil {
+		t.Fatalf("cannot write %d bytes to %s: %s", len(data), address, err)
+	}
+	if n != len(d) {
```
Comment (rtm0) [inline apptest/client.go:92]: I guess we want to get the lentgh of the byte slice instead of string because if the string contains unicode symbols that are non-ascii the string length will be different from byte length.

File: apptest/tests/ingestprotocols_test.go [hunk 3]
```diff
@@ -73,22 +79,105 @@ func TestSingleIngestionProtocols(t *testing.T) {
 		},
 	})
 
+	// OpenTSDB HTTP write format
+	sut.OpenTSDBHTTPImport(t, []string{
+		`{"metric":"opentsdbimport.foo","value":45.34, "timestamp": "1707123457"}`,
```
Comment (rtm0) [inline apptest/tests/ingestprotocols_test.go:84]: Here and everywhere else: Please add a comment which date the timestamps correspond to.

File: apptest/tests/special_query_regression_test.go [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+	"os"
+	"testing"
+)
+
+// TestSpecialQueryRegression is used to test queries that have experienced issues for specific data sets.
+// These test cases were migrated from `app/victoria-metrics/main_test.go`.
+// Most of these cases are based on user feedback. Refer to the corresponding GitHub issue for details on each case.
+//
+// To improve performance, it will handle ingestion and flushing at one location and then evaluate each query.
+func TestSpecialQueryRegression(t *testing.T) {
```
Comment (rtm0) [inline apptest/tests/special_query_regression_test.go:16]: It is important to follow the following naming convention for the test names: a test name must start either with TestSingle or TestCluster. That way we can distinguish which tests are run for vmsingle (master branch) and vmluster (cluster branch). If you look at the integration-test build rule in the Makefile you will notice that it skips either TestSingle or TestCluster depending on the branch. 

So in case of this test you need to change it to start with `TestSingle`.

Also, since apptest supports testing clusters, this and other tests that you've added can also be written for the cluster. See other existing tests for example (such as key_concenpts_test.go). Not for this PR though. Could be a follow up PR.

**Review by rtm0 — CHANGES_REQUESTED** [review]
Nice! The new tests appear much more conscise than the original ones.

Few remarks though.

File: apptest/vminsert.go [hunk 5]
```diff
@@ -53,6 +58,8 @@ func StartVminsert(instance string, flags []string, cli *Client) (*Vminsert, err
 		defaultFlags: map[string]string{
 			"-httpListenAddr":          "127.0.0.1:0",
 			"-clusternativeListenAddr": "127.0.0.1:0",
+			"-graphiteListenAddr":      ":2003",
+			"-opentsdbListenAddr":      "127.0.0.1:4242",
```
[the remark this answers is not in the document]
Comment (jiekun) [reply apptest/vminsert.go:62]: Updated

File: apptest/vminsert.go [hunk 6]
```diff
@@ -87,8 +96,46 @@ func (app *Vminsert) InfluxWrite(t *testing.T, records []string, opts QueryOpts)
...
+	app.cli.Write(t, app.graphiteWriteURL, records)
+}
+
+// CSVImport is a test helper function that inserts a
+// collection of records in CSV format for the given
+// tenant by sending an HTTP POST request to
+// prometheus/api/v1/import/csv vminsert endpoint.
+//
+// See https://docs.victoriametrics.com/cluster-victoriametrics/#url-format
+func (app *Vminsert) CSVImport(t *testing.T, records []string, opts QueryOpts) {
```
[the remark this answers is not in the document]
Comment (jiekun) [reply apptest/vminsert.go:129]: Renamed

File: apptest/tests/special_query_regression_test.go (hunk 4 above)
Comment (jiekun) [reply apptest/tests/special_query_regression_test.go]: Renamed to `TestSingleSpecialQueryRegression`

File: apptest/tests/ingestprotocols_test.go (hunk 3 above)
Comment (jiekun) [reply apptest/tests/ingestprotocols_test.go:84]: Done

File: apptest/client.go (hunk 2 above)
Comment (jiekun) [reply apptest/client.go]: Thanks! fixed.

File: apptest/vmsingle.go [hunk 7]
```diff
@@ -69,6 +75,8 @@ func StartVmsingle(instance string, flags []string, cli *Client) (*Vmsingle, err
 		forceFlushURL:                      fmt.Sprintf("http://%s/internal/force_flush", stderrExtracts[1]),
 		forceMergeURL:                      fmt.Sprintf("http://%s/internal/force_merge", stderrExtracts[1]),
 		influxLineWriteURL:                 fmt.Sprintf("http://%s/influx/write", stderrExtracts[1]),
+		graphiteWriteURL:                   stderrExtracts[2],
+		openTSDBHTTPURL:                    fmt.Sprintf("http://%s/api/put", stderrExtracts[3]),
```
[the remark this answers is not in the document]
Comment (jiekun) [reply apptest/vmsingle.go]: PTAL

File: apptest/vminsert.go [hunk 8]
```diff
@@ -87,8 +96,46 @@ func (app *Vminsert) InfluxWrite(t *testing.T, records []string, opts QueryOpts)
...
+
+	url := fmt.Sprintf("http://%s/insert/%s/prometheus/api/v1/import/prometheus", app.httpListenAddr, opts.getTenant())
+	uv := opts.asURLValues()
+	uvs := uv.Encode()
+	if len(uvs) > 0 {
+		url += "?" + uvs
+	}
+	data := []byte(strings.Join(records, "\n"))
 	app.sendBlocking(t, len(records), func() {
 		_, statusCode := app.cli.Post(t, url, "text/plain", data)
```
[the remark this answers is not in the document]
Comment (jiekun) [reply apptest/vminsert.go:139]: Checked it again but still wonder why `PostForm`?

File: apptest/vminsert.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jiekun) [reply apptest/vminsert.go]: done

**Comment by jiekun** [comment]
All updated. PTAL again. Thanks!

File: apptest/tests/special_query_regression_test.go [hunk 9]
```diff
@@ -0,0 +1,477 @@
...
+
+import (
+	at "github.com/VictoriaMetrics/VictoriaMetrics/apptest"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"os"
+	"testing"
+)
+
+// TestSpecialQueryRegression is used to test queries that have experienced issues for specific data sets.
```
Comment (rtm0) [inline apptest/tests/special_query_regression_test.go:11]: TestSingleSpecialQueryRegression

File: apptest/tests/special_query_regression_test.go [hunk 10]
```diff
@@ -0,0 +1,477 @@
...
+func TestSingleSpecialQueryRegression(t *testing.T) {
+	os.RemoveAll(t.Name())
+	tc := at.NewTestCase(t)
+	defer tc.Stop()
+
+	sut := tc.MustStartDefaultVmsingle()
+
+	// prometheus
+	setupPrometheusImport(t, sut)
+	testCaseSensetiveRegex(t, tc, sut)
```
Comment (rtm0) [inline apptest/tests/special_query_regression_test.go:25]: Sensetive -> Sensitive

Also just pass `tc` to the function. Inside that function, you can get t from tc like this: `t := tc.T()`. Same for other functions.

File: apptest/tests/special_query_regression_test.go [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+		`GenBearTemp{db="TenMinute",Park="3",TurbineType="V112"} 10 1707123456700`, // 2024-02-05T08:57:36.700Z
+		`GenBearTemp{db="TenMinute",Park="4",TurbineType="V112"} 10 1707123456700`, // 2024-02-05T08:57:36.700Z
+	}, at.QueryOpts{})
+
+	sut.ForceFlush(t)
+}
+
+// setupGraphiteImport import data for each test cases using Graphite TCP conn.
+func setupGraphiteImport(t *testing.T, tc *at.TestCase, sut *at.Vmsingle) {
+	rowInserted := sut.ServesMetrics.GetIntMetric(t, `vm_rows_inserted_total{type="graphite"}`)
```
Comment (rtm0) [inline apptest/tests/special_query_regression_test.go:79]: I believe you should be able to just `sut.GetIntMetric`

File: apptest/tests/special_query_regression_test.go [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"os"
+	"testing"
+)
+
+// TestSpecialQueryRegression is used to test queries that have experienced issues for specific data sets.
+// These test cases were migrated from `app/victoria-metrics/main_test.go`.
+// Most of these cases are based on user feedback. Refer to the corresponding GitHub issue for details on each case.
+//
+// To improve performance, it will handle ingestion and flushing at one location and then evaluate each query.
```
Comment (rtm0) [inline apptest/tests/special_query_regression_test.go:15]: I think we can sacrifice some performace in favor of readability. Let's place the data ingestion and retrieval in the same function so it we don't have to navigate back and forth when debugging.

File: apptest/tests/special_query_regression_test.go [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+		Got: func() any {
+			return (sut.ServesMetrics.GetIntMetric(t, `vm_rows_inserted_total{type="graphite"}`) - rowInserted) >= 23
+		},
+		Want: true,
+	})
+
+	sut.ForceFlush(t)
+}
+
+func testCaseSensetiveRegex(t *testing.T, tc *at.TestCase, sut *at.Vmsingle) {
```
Comment (rtm0) [inline apptest/tests/special_query_regression_test.go:128]: Sensitive

File: apptest/vminsert.go [hunk 14]
```diff
@@ -87,8 +96,45 @@ func (app *Vminsert) InfluxWrite(t *testing.T, records []string, opts QueryOpts)
...
+
+// PrometheusAPIV1ImportCSV is a test helper function that inserts a collection
+// of records in CSV format for the given tenant by sending an HTTP POST
+// request to prometheus/api/v1/import/csv vminsert endpoint.
+//
+// See https://docs.victoriametrics.com/cluster-victoriametrics/#url-format
+func (app *Vminsert) PrometheusAPIV1ImportCSV(t *testing.T, records []string, opts QueryOpts) {
+	t.Helper()
+
+	url := fmt.Sprintf("http://%s/insert/%s/prometheus/api/v1/import/prometheus", app.httpListenAddr, opts.getTenant())
```
Comment (rtm0) [inline apptest/vminsert.go:131]: Looks like the URL misses `csv`

File: apptest/tests/special_query_regression_test.go (hunk 12 above)
Comment (jiekun) [reply apptest/tests/special_query_regression_test.go]: I moved them to each test cases. Running locally and it costs extra 2-4 seconds. It should be fine for now tho.

**Comment by jiekun** [comment]
Thank you @rtm0 ! All updated. PTAL again when you have some free time.

**Review by rtm0 — APPROVED** [review]
LGTM, thanks!

**Comment by rtm0** [comment]
Once this PR is merged, please add the same set of tests for cluster. You don't have to duplicate them though. They can share the same test cases with vmsingle. You just need to setup a cluster and run the existing test cases against it. See other tests for example.

**Comment by rtm0** [comment]
@jiekun Since you mentioned you wanted to add the cluster version in this PR I will postpone merging until then.

**Comment by jiekun** [comment]
@rtm0 Thank you.

I'm proposing the following change to the interface `PrometheusWriteQuerier`:
```go
// MetricsWriterPrometheusQuerier encompasses the methods for writing, flushing and
// querying the data.
type MetricsWriterPrometheusQuerier interface {
	PrometheusWriter
	PrometheusQuerier
	GraphiteWriter
	OpenTSDBWriter
	StorageFlusher
	StorageMerger
}
```

So previously it was designed to handle everything for the Prometheus API. If we want to add support for additional protocols, we may need different interfaces in both definition and test code, which seems complicated. What do you think about the current change, or do you have any other suggested approaches?

**Comment by rtm0** [comment]
> I'm proposing the following change to the interface `PrometheusWriteQuerier`:

Yes, the interfaces need to be extended with the new API calls that you've added. Let's maybe just add the corresponding methods to the existing PrometheusWriter interface and rename it so that the name is more generic. What I mean is:

```go
type Writer interface {
  PrometheusAPIV1*()
  GraphiteWrite()
  OpenTSDBAPIPut()
} 
```

This seems to be the only change needed. If we need more ways to query data, then those can also be added to the PrometheusQuerier interface and the interface could be renamed to something like `Querier`.

These two interfaces PrometheusQuerier and `PrometheusWriter` were named so because we only used Prometheus API when we created these interfaces. And I am not sure whether we want to have different interfaces for different APIs. I'd just put all query APIs under one interface and all writer apis under one interface to keep things simple.

File: apptest/model.go [hunk 15]
```diff
@@ -23,20 +23,17 @@ type PrometheusQuerier interface {
 	PrometheusAPIV1Series(t *testing.T, matchQuery string, opts QueryOpts) *PrometheusAPIV1SeriesResponse
 }
 
-// PrometheusWriter contains methods available to Prometheus-like HTTP API for Writing new data
-type PrometheusWriter interface {
+// Writer contains methods for Writing new data
```
Comment (rtm0) [inline apptest/model.go:26]: Writing -> writing

File: apptest/testcase.go [hunk 16]
```diff
@@ -267,8 +267,8 @@ func (tc *TestCase) StopApp(instance string) {
 	}
 }
 
-// StopPrometheusWriteQuerier stop all apps that are a part of the pwq.
-func (tc *TestCase) StopPrometheusWriteQuerier(pwq PrometheusWriteQuerier) {
+// StopMetricsWriterQuerier stop all apps that are a part of the pwq.
+func (tc *TestCase) StopMetricsWriterQuerier(pwq PrometheusWriteQuerier) {
```
Comment (rtm0) [inline apptest/testcase.go:271]: Please rename to StopPrometheusWriteQuerier since the interface is still named PrometheusWriteQuerier

File: apptest/tests/special_query_regression_test.go [hunk 17]
```diff
@@ -0,0 +1,571 @@
...
+	testDuplicateLabel(tc, sut)
+	testTooBigLookbehindWindow(tc, sut)
+	testMatchSeries(tc, sut)
+
+	// graphite
+	testComparisonNotInfNotNan(tc, sut)
+	testEmptyLabelMatch(tc, sut)
+	testMaxLookbehind(tc, sut)
+	testNonNanAsMissingData(tc, sut)
+	testSubqueryAggregation(tc, sut)
```
Comment (rtm0) [inline apptest/tests/special_query_regression_test.go:58]: Please extract these into a separate function and pass sut to it in both TestSingle and TestCluster. This should work because both vmsingle and vmcluster implement that interface.

**Review by rtm0 — CHANGES_REQUESTED** [review]
Almost there. A few comments to address.

**Review by rtm0 — APPROVED** [review]
LGTM, thanks!

**Review by f41gh7 — APPROVED** [review]
LGTM
